src/dynamic_honey_badger.rs

```diff
@@ -0,0 +1,449 @@
+    vote_counts: HashMap<Change<NodeUid>, usize>,
+    /// The `HoneyBadger` instance with the current set of nodes.
+    honey_badger: HoneyBadger<Transaction<Tx, NodeUid>, NodeUid>,
+    /// The current key generation process.
+    key_gen: Option<SyncKeyGen<NodeUid>>,
+    /// A queue for messages from future epochs that cannot be handled yet.
+    incoming_queue: Vec<(u64, NodeUid, honey_badger::Message<NodeUid>)>,
+    /// The messages that need to be sent to other nodes.
+    messages: MessageQueue<NodeUid>,
+    /// The outputs from completed epochs.
+    output: VecDeque<Batch<Tx, NodeUid>>,
+}
+
+impl<Tx, NodeUid> DistAlgorithm for DynamicHoneyBadger<Tx, NodeUid>
+where
+    Tx: Eq + Serialize + for<'r> Deserialize<'r> + Debug + Hash,
+    NodeUid: Eq + Ord + Clone + Serialize + for<'r> Deserialize<'r> + Debug + Hash,
+{
+    type NodeUid = NodeUid;
+    type Input = Input<Tx, NodeUid>;
+    type Output = Batch<Tx, NodeUid>;
+    type Message = Message<NodeUid>;
+    type Error = Error;
+
+    fn input(&mut self, input: Self::Input) -> Result<()> {
+        let tx = self.convert_transaction(input);
+        self.honey_badger.input(tx)?;
+        self.process_output()
+    }
+
+    fn handle_message(&mut self, sender_id: &NodeUid, message: Self::Message) -> Result<()> {
+        if !self.netinfo.all_uids().contains(sender_id) {
+            return Err(ErrorKind::UnknownSender.into());
+        }
+        match message {
+            Message::HoneyBadger(start_epoch, hb_msg) => {
+                self.handle_honey_badger_message(sender_id, start_epoch, hb_msg)
+            }
+        }
+    }
+
+    fn next_message(&mut self) -> Option<TargetedMessage<Self::Message, NodeUid>> {
+        self.messages.pop_front()
+    }
+
+    fn next_output(&mut self) -> Option<Self::Output> {
+        self.output.pop_front()
+    }
+
+    fn terminated(&self) -> bool {
+        false
+    }
+
+    fn our_id(&self) -> &NodeUid {
+        self.netinfo.our_uid()
+    }
+}
+
+impl<Tx, NodeUid> DynamicHoneyBadger<Tx, NodeUid>
+where
+    Tx: Eq + Serialize + for<'r> Deserialize<'r> + Debug + Hash,
+    NodeUid: Eq + Ord + Clone + Debug + Serialize + for<'r> Deserialize<'r> + Hash,
+{
+    /// Returns a new instance with the given parameters, starting at epoch `0`.
+    pub fn new(netinfo: NetworkInfo<NodeUid>, batch_size: usize) -> Result<Self> {
+        let honey_badger = HoneyBadger::new(Rc::new(netinfo.clone()), batch_size, None)?;
+        let dyn_hb = DynamicHoneyBadger {
+            netinfo,
+            batch_size,
+            start_epoch: 0,
+            votes: BTreeMap::new(),
+            vote_counts: HashMap::new(),
+            honey_badger,
+            key_gen: None,
+            incoming_queue: Vec::new(),
+            messages: MessageQueue(VecDeque::new()),
+            output: VecDeque::new(),
+        };
+        Ok(dyn_hb)
+    }
+
+    fn convert_transaction(&self, input: Input<Tx, NodeUid>) -> Transaction<Tx, NodeUid> {
```

Comment:
I think that a better abstraction here would be to implement a conversion trait instead of using a method (because `&self` isn't being used):

```rust
impl<Tx, NodeUid> From<Input<Tx, NodeUid>> for Transaction<Tx, NodeUid> {
    match input {
        Input::User(tx) => Transaction::User(tx),
        Input::Change(change) => Transaction::Change(change),
    }
}
```

Then converting from `Input -> Transaction` could be written:

```rust
let tx = input.into();
```

or, if the type parameters are necessary:

```rust
let tx: Transaction<Self::Input, Self::NodeUid> = input.into();
```

Reply:
You're right, I'll change it. (When I originally wrote the method, I still needed `self`.)